main.go

```diff
@@ -8,64 +8,60 @@ import (
 )
 
 // validateIPv4 function to check a given string is valid ipv4 or not.
-func validateIPv4(line string) bool {
+func validateIPv4(line string) (bool, error) {
 	ipv4, err := regexp.Compile(`^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])$`)
 	if err != nil {
-		log.Fatal(err)
+		return false, err
 	}
 	if ipv4.MatchString(line) {
-		return true
+		return true, err
 	} else {
-		return false
+		return false, err
 	}
 }
 
 // validateIPv6 function to check a given string is valid ipv6 or not.
-func validateIPv6(line string) bool {
+func validateIPv6(line string) (bool, error) {
 	ipv6, err := regexp.Compile(`^((([0-9a-fA-F]){1,4})\:){7}([0-9a-fA-F]){1,4}$`)
 	if err != nil {
-		log.Fatal(err)
+		return false, err
 	}
 	if ipv6.MatchString(line) {
-		return true
+		return true, err
 	} else {
-		return false
+		return false, err
 	}
 }
 
 func ip(w http.ResponseWriter, r *http.Request) {
 
-	if r.URL.Path != "/" {
-		http.Error(w, "404 not found.", http.StatusNotFound)
-		return
+	inputIP := r.FormValue("ip")
+
+	if inputIP == "" {
+		http.Error(w, "missing value", http.StatusBadRequest)
 	}
 
-	switch r.Method {
-	case "GET":
-		http.ServeFile(w, r, "form.html")
-	case "POST":
-		if err := r.ParseForm(); err != nil {
-			fmt.Fprintf(w, "ParseForm() err: %v", err)
-			return
-		}
-		inputIP := r.FormValue("IP")
-		if validateIPv4(inputIP) {
-			w.Write([]byte(`<div style = "display:flex; justify-content:center; align-items:center; height:100vh;"><h2>valid IPv4 </h2></div>`))
-		} else if validateIPv6(inputIP) {
-			w.Write([]byte(`<div style = "display:flex; justify-content:center; align-items:center; height:100vh;"><h2>valid IPv6 </h2></div>`))
-		} else {
-			w.Write([]byte(`<div style = "display:flex; justify-content:center; align-items:center; height:100vh;"><h2>invalid IP </h2></div>`))
-		}
-	default:
-		w.Write([]byte(`<div style = "display:flex; justify-content:center; align-items:center; height:100vh;"><h2>Only GET and POST methods are supported. </h2></div>`))
+	ipv4, ipv4err := validateIPv4(inputIP)
+	ipv6, ipv6err := validateIPv6(inputIP)
+
+	if ipv4err != nil {
+		http.Error(w, ipv4err.Error(), http.StatusBadRequest)
+	} else if ipv4 {
+		fmt.Fprintln(w, "validIPv4")
+	} else if ipv6err != nil {
+		http.Error(w, ipv6err.Error(), http.StatusBadRequest)
+	} else if ipv6 {
+		fmt.Fprintln(w, "validIPv6")
+	} else {
+		fmt.Fprintln(w, "invalidIP")
 	}
```

Comment:
**Suggestion:** In the ip function, you are calling both validateIPv4 and validateIPv6 functions regardless of whether the IP is valid or not. You can improve performance by returning as soon as a valid IP is found.
```suggestion
 ipv4, ipv4err := validateIPv4(inputIP)
 if ipv4err != nil {
 	http.Error(w, ipv4err.Error(), http.StatusBadRequest)
 	return
 }
 if ipv4 {
 	fmt.Fprintln(w, "validIPv4")
 	return
 }
 ipv6, ipv6err := validateIPv6(inputIP)
 if ipv6err != nil {
 	http.Error(w, ipv6err.Error(), http.StatusBadRequest)
 	return
 }
 if ipv6 {
 	fmt.Fprintln(w, "validIPv6")
 	return
 }
 fmt.Fprintln(w, "invalidIP")
```